R/join_keys-extract.R

```diff
@@ -0,0 +1,107 @@
+#' @rdname join_keys
+#'
+#' @details
+#' - `[.join_keys` can be used to return a subset of relationship pairs. It will
+#' retrieve the primary keys of the selected elements and its parents (along with)
+#' the relationship keys between the selected elements and their parents.
+#'
+#' @param i index specifying elements to extract or replace. Index should be a
+#' a character vector, but it can also take numeric, logical, `NULL` or missing.
+#'
+#' @param j index specifying elements to extract or replace. Index should be a
+#' a character vector, but it can also take numeric, logical, `NULL` or missing.
+#'
+#' @export
+#'
+#' @examples
+#'
+#' # Getter for join_keys ----
+#'
+#' jk <- join_keys(
+#'   join_key("ds1", "ds1", "primary-key-1"),
+#'   join_key("ds2", "ds2", "primary-key-2"),
+#'   join_key("ds3", "ds3", "primary-key-3"),
+#'   join_key("ds2", "ds1", "foreign-key-2-1"),
+#'   join_key("ds3", "ds1", "foregin-key-3-1")
+#' )
+#'
+#' jk["ds1"]
+#' jk[1:2]
+#' jk[c("ds1", "ds2")]
+#' jk["ds1", "ds2"]
+`[.join_keys` <- function(x, i, j) {
+  if (missing(i)) {
+    return(x)
+  }
+
+  if (is.null(i)) {
+    return(join_keys()) # replicate base R
+  }
+
+  if (!missing(j)) {
+    checkmate::assert(
+      combine = "or",
+      checkmate::check_string(i),
+      checkmate::check_integerish(i, len = 1),
+      checkmate::check_logical(i, len = length(x))
+    )
+    checkmate::assert(
+      combine = "or",
+      checkmate::check_string(j),
+      checkmate::check_integerish(j, len = 1),
+      checkmate::check_logical(j, len = length(x))
+    )
+
+    subset_x <- update_keys_given_parents(x)[union(i, j)]
```

Comment:
A small performance boost

```suggestion
    subset_x <- update_keys_given_parents(x[union(i, j)])
```

Reply:
I'll try. Thanks!